records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,349 @@
+            '"job_names": ["{job_name}"]}}'
+        ).format(project_id=project_id, job_name=job_name)
+
+        done = False
+        logger.info("Awaiting completion of Google Cloud Platform Data Transfer Service job...")
+        while not done:
+            result = storagetransfer.transferOperations().list(
+                name="transferOperations",
+                filter=filterString).execute()
+            logger.debug('Result of transferOperations/list: {}'.format(
+                json.dumps(result, indent=4, sort_keys=True)))
+            if result != {}:
+                done = result['operations'][0]['metadata']['status'] != 'IN_PROGRESS'
+            time.sleep(10)
+        logger.info("Google Cloud Platform Data Transfer Service job complete.")
+
+    def try_swapping_bucket_path(self,
+                                 target_bucket: Union['GCSDirectoryUrl', 'S3DirectoryUrl'],
+                                 bucket_to_steal_path_from: Union['GCSDirectoryUrl',
+                                                                  'S3DirectoryUrl']) ->\
+            Optional[Union['GCSDirectoryUrl', 'S3DirectoryUrl']]:
+        optimized_directory =\
+            target_bucket.directory_in_this_bucket(bucket_to_steal_path_from.key)
+        if not optimized_directory.empty():
+            logger.info(f"{optimized_directory} is not empty")
+            return None
+
+        if not optimized_directory.writable():
+            logger.info(f"{optimized_directory} is not writable")
+            return None
+
+        return optimized_directory
+
+    @contextmanager
+    def _optimize_temp_locations_for_gcp_data_transfer(self,
+                                                       temp_first_loc: 'S3DirectoryUrl',
+                                                       temp_second_loc: 'GCSDirectoryUrl') ->\
+            Iterator[Tuple[BaseDirectoryUrl, BaseDirectoryUrl]]:
+        #
+        # GCP data transfer is great, but has the limitations:
+        #
+        # 1) It only works from GCS -> GCS or S3 -> GCS - like the
+        # roach motel, you can check in but you can't check out.
+        #
+        # 2) You can't specify a different destination location
+        # directory in the GCS bucket than in your source bucket
+        #
+        # So let's make sure that if at all possible, we use the same
+        # directory.
+        optimized_temp_first_loc = self.try_swapping_bucket_path(temp_first_loc,
+                                                                      temp_second_loc)
+        if optimized_temp_first_loc is not None:
+            # TODO: Delete after
+            yield (optimized_temp_first_loc, temp_second_loc)
+            return
+        optimized_temp_second_loc = self.try_swapping_bucket_path(temp_second_loc,
+                                                                    temp_first_loc)
+        if optimized_temp_second_loc is not None:
+            # TODO: Delete after
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker